spec/abstract-jobs.html

```diff
@@ -1,6 +1,63 @@
 <emu-clause id="sec-abstract-jobs">
   <h1> Abstract Jobs </h1>
 
+  <emu-clause id="sec-weakref-processing-model">
+    <h1>Processing model of WeakRefs and FinalizationGroups</h1>
+
+    <p>
+      This specification does not make any guarantees about garbage collection.
+      Objects which are unreachable from ECMAScript may be released after long periods
+      of time, or never at all. For this reason, this specification uses the term
+      "may" when describing behavior triggered by garbage collection.
+    </p>
+
+    <p>
+      At any given time, if all references to a ECMAScript object are from a
+      [[Target]] field or internal slot, inlcuding the agent's [[KeptAlive]] field,
+      the implementation may replace the reference to the object in all such
+      [[Target]] fields or internal slots with the value ~empty~.
+    </p>
+
+    <p>
+      The semantics of WeakRefs and FinalizationGroups is based on two
+      operations which happen at particular points in time:
+    </p>
+
+    <ul>
+      <li>
+        When WeakRef.prototype.deref is called, the referent (if it's not
+        already dead) is kept alive so that subsequent, synchronous accesses also
+        return the object. This list is reset when synchronous work is done using
+        the ClearKeptObjects abstract operation.
+      </li>
+
+      <li>
+        When an object which is registered with a FinalizationGroup becomes
+        unreachable, a call of the FinalizationGroup's cleanup callback may
+        eventually be made, after synchronous JavaScript execution completes.
+        The FinalizationGroup cleanup is performed with the
+        FinalizationGroupCleanup abstract operation.
+      </li>
+    </ul>
+
+    <p>
+      Neither of these actions (ClearKeptObjects or FinalizationGroupCleanup)
```

Comment:
It seems like this text is going out of its way to avoid just calling these Jobs (aka turns), and leaving it to the host to determine when these Jobs interleave with other Jobs. The language seems to lose the clarity of Jobs as a unit of atomicity.

Reply:
I don't think we have the machinery in ECMAScript to specify that these are scheduled the way that they should be, to not interrupt a higher priority sequence of Promise reactions.

W could specify this in terms of Jobs, but this would force HTML to return to making another "willful violation" in order to specify how they are ordered with respect to other Jobs (since the HTML event loop isn't in terms of ECMAScript Jobs).

We know from experience that such "willful violations" are confusing to implementers. In particular, I'm concerned that someone could see this Job language and conclude that these actions *are* interspersed with Promise reactions.

Reply:
The text that you wrote already allows these to be interleaved with promise jobs. 

The html spec not being written in terms of ES jobs seems like a bug. Why is any willful violation necessary?

Reply:
That's true, it does allow it to be interleaved with Promise Jobs, but it also allows embedder specifications to precisely say that they are not interleaved. If we said it in terms of Jobs, then it's really difficult to specify the ordering--RunJobs says that the ordering is implementation-defined (not embedder-defined) and is generally hard to plug into while providing very little structure for itself.

I wrote a presentation explaining the willful violation and how we can avoid it: [slides](https://docs.google.com/presentation/d/1LLIIwZyQgm6rbAyWbRjFFSNKlA2DPqslpN8sVKjchjE/edit#slide=id.p). I was going to present this at the March 2019 TC39 meeting, but delayed it as there's a bit of ongoing refactoring in the event loop at the HTML level. At this point, I'm thinking of presenting basically the same thing in June.

If it's really important to you, I can rewrite this in terms of ES Jobs, but then I'll be forced to broaden the willful violation in HTML. That doesn't make me happy.

Reply:
I share the goal of making the JS spec appropriate for non-browser hosts. I'd like to consider other possibilities for making the specification precise for these various scenarios. I'm glad we agree that it's important that we fix the mismatch.

I think, at this point, we can agree that it's an (important!) editorial question we're talking about, and not a normative one. Can we work towards resolving this question during Stage 3/while reviewing that presentation, unblocking WeakRefs from advancing?

The big semantic point that this patch tries to make, together with the HTML PR, is to say, "[to the extent that we can say this] WeakRef deref() holds the object live until until after all the Promise reactions run, and the FinalizationGroup callbacks are not interspersed in Promise reactions; the garbage collector may launch FinalizationGroup callbacks even if JS is idle".

The parenthetical at the beginning is because JS currently leaves so much up to hosts, so it's hard to say anything about Promise reactions happening before anything else in particular. However, hosts have so much going on in them, that it's hard to figure out how to say much more at the JS level. Maybe we should have a declarative statement somewhere explaining the prioritization, to be specified imperatively in more detail by the host.

Reply:
> WeakRef deref() holds the object live until until after all the Promise reactions run

What is the reasoning behind this requirement? I thought the goal was to prevent collection of those targets during the current *synchronous* code.\
The note for `KeepDuringJob` says:

> This may be abstractly implemented as a Set in the current Job that becomes unreachable when the Job ends because the Job becomes unreachable, or as a Set in the current Agent that gets cleared at the end of each Job.

Wouldn't that suggestion allow collection before the Promise jobs queue is emptied?

```javascript
let wr = new WeakRef({});
Promise.resolve().then(() => {
  gc();
  console.log(wr.deref()); // is undefined allowed here?
});
```

Reply:
No, undefined is not allowed there. The "Job" terminology is ambiguous, so this PR switches to using host hooks that define things more precisely. The kept set is not cleared except by a new task, which will run only after a microtask checkpoint.